### ISSUE TITLE
Proposal: Compose Transform

### ISSUE THREAD

**Issue by Jeffail** [vectordotdev/vector#1653]
I'd like to outline a new proposal, taking concepts from both https://github.com/timberio/vector/issues/1447 and https://github.com/timberio/vector/issues/1061. To summarise our goal here we want to solve two main use cases:

1. I want to be able to define a chain of transforms as a list, without the need to manually hook them using the `inputs` field. This will make it easier to read and write simple pipelines, and less likely to make mistakes when adding, removing or re-ordering the transforms.

2. I want to be able to write and test complex topologies of transforms in isolation, and then insert it within a larger production topology configuration. This will make it easier to maintain large projects.

That's it. Based on use cases and feedback we've gathered so far I'm specifically leaving out sources and sinks, as I think in practice the main issue is being able to compose transforms. In terms of making sources and sinks easier to hook up we have options but I'd like to keep them separate from this proposal.

## The Proposal

In order to solve use case 1 we introduce a new transform `compose` which, as part of its definition, allows you to configure a list of inline transforms:

```toml
[transforms.foo_pipe]
  inputs = [ "TODO" ]
  type = "compose"

  [[transforms.foo_pipe.children]]
    type = "foo"

  [[transforms.foo_pipe.children]]
    type = "baz"
```

Inserting a transform `bar` is then as simple as pasting it in:

```diff
 [transforms.foo_pipe]
   inputs = [ "TODO" ]
   type = "compose"

   [[transforms.foo_pipe.children]]
     type = "foo"

+  [[transforms.foo_pipe.children]]
+    type = "bar"

   [[transforms.foo_pipe.children]]
     type = "baz"
```

These children transforms are not necessarily anonymous as we can identify them by index (`foo_pipe_child_0`, etc). However, I think in practice metrics at the individual transform level will not be mission critical in any case. More importantly these inline transforms do not specify an `inputs` list and they do not need to, the `compose` transform takes ownership and responsibility for hooking up all child transforms.

This solves use case 1 as it provides a clean and concise way to specify a chain of transforms, where it's much easier and hazard free to insert and re-order them. However, this doesn't yet solve use case 2 as it still doesn't provide component reusability. In order to solve that we can extend the `compose` transform to also support transform references in place of inline transforms:

```toml
[transforms.foo]
  inputs = [ "TODO" ]
  type = "foo"

[transforms.bar]
  inputs = [ "TODO" ]
  type = "bar"

[transforms.baz]
  inputs = [ "TODO" ]
  type = "baz"

[transforms.foo_pipe]
  inputs = [ "TODO" ]
  type = "compose"
  children = [
  	"foo", "bar", "baz"
  ]
```

When a compose transform references transforms in this way it creates a _copy_ of it rather than connecting to the real global transform. This avoids odd behaviors where isolated sections of a topology might bleed into each other (as we saw with the pipelines proposal). We can make this behavior clear in the documentation for the compose transform.

When using composed transforms it's likely that each referenced transform will _only_ be used this way, in which case the `inputs` parameter is redundant and can be removed. We must remember to update our linting behavior such that this case is not considered an error.

## Problems

TOML becomes a nuisance when dealing with deeply nested structures, placing transforms inline like this might start looking ugly.

Another consideration with `compose` transforms is that over reliance on them, combined with poorly structured pipelines, can result in duplicated effort. For example:

```toml
[transforms.do_some_structuring]
  type = "compose"
  children = [ "foo", "bar" ]

[transforms.compose_filter_1]
  inputs = [ "TODO" ]
  type = "compose"
  children = [ "do_some_structuring", "filter_1" ]

[transforms.compose_filter_2]
  inputs = [ "TODO" ]
  type = "compose"
  children = [ "do_some_structuring", "filter_2" ]
```

In the the above example the effort of transforms "foo" and "bar" is duplicated. A more efficient structure would be:

```toml
[transforms.do_some_structuring]
  inputs = [ "TODO" ]
  type = "compose"
  children = [ "foo", "bar" ]

[transforms.filter_1]
  inputs = [ "do_some_structuring" ]
  type = "filter"

[transforms.filter_2]
  inputs = [ "do_some_structuring" ]
  type = "filter"
```

A linting rule might be able to capture this, but the implementation would need to be exhaustive and the error messages will need to provide heavy guidance in order to be useful.

## Closing thoughts

I believe this proposal is a nice complement to our existing format. It doesn't introduce any dramatically new paradigms, and is therefore easy to document and simple to learn. It can be a great asset for teams working on advanced configs and makes slapping simple pipelines together easier.

This proposal does come with caveats. However, given we're only really introducing one feature (a new transform) it would be pain-free to back out of this feature if we later decide it isn't a good fit.

## Possible further steps

Expand beyond linear lists of transforms so that topologies with branching (using filters) can be composed.

**Comment by binarylogic** [comment]
Nice write up! I like how you're thinking about this. Fitting this into the transform paradigms helps to keep things consistent and discoverable. On the other hand, I think it fails to address the problems mentioned in https://github.com/timberio/vector/issues/1327#issuecomment-569018428:

> My biggest pains with the current format are :
>
> 1. the graph is completely implicit and hard to decode as you must read every inputs fields
> 2. I don't like the "array of tables" TOML syntax (I don't like list of objects in general and I think we are generally best served with plain objects/tables)
> 3. I can't split the configuration in several files

Your above proposal is a good solution to #1061 but not #1447 and the above comments (in my opinion). If were to think of this as a solution for #1061 _only_ then I think it makes more sense, but then we're still left solving #1447. And that's the problem I have. I think, if done correctly, #1447 can solve both in a simpler manner.

To use your examples above:

```toml
# vector/config/foo_pipe.toml

pipelines.foo_pipe = ["foo", "bar", "baz"]

[transforms.foo]
  type = "foo"

[transforms.bar]
  type = "bar"

[transforms.baz]
  type = "baz"
```

In my opinion, this is much simpler and easier to read. But we're still left with a couple of problems:

1. The edge case mentioned in #1447.
2. Exposing transforms that shouldn't be used individually. For example, the author of this pipeline might want to ensure that only the pipeline itself can be referenced, not it's components.

I'll address both of these separately:

### 1. The copy/reference edge case in #1447 

Above you mention:

> When a compose transform references transforms in this way it creates a copy of it rather than connecting to the real global transform.

I like this. Copying makes sense and it's something we can make very clear in the documentation. Additionally, when referencing that transform we can use an explicit ID that includes the pipeline reference. For example, the `bar` transform above would become `foo_pipe.bar` and that's the ID that would be used in logs, metrics labels, etc.

### 2. Exposing transforms that shouldn't be used directly

To address #1061 we can allow for private transforms. Perhaps prefixing the name with `_` would ensure that the transform could not be used direct and _only_ on pipelines. For example:

```toml
# vector/config/foo_pipe.toml

pipelines.foo_pipe = ["_foo", "_bar", "_baz"]

[transforms._foo]
  type = "foo"

[transforms._bar]
  type = "bar"

[transforms._baz]
  type = "baz"
```

The author of this pipeline is not explicitly stating that the pipeline should be used as a whole. This effectively turns the pipeline into a macro/reusable transform. It also prevents other collaborators from misusing the transforms and gives the author of this pipeline freedom to change the pipeline however they see fit.

---

To sum up my thoughts:

- [ ] Add a new `pipelines.*` directive
- [ ] Make `inputs` optional since topologies can now be connected through pipelines.
- [ ] Warn if a component is not reference via `inputs` or within `pipelines`
- [ ] Allow for private component IDs by prefixing them with `_`. Using an ID prefixed with `_` in `inputs` would error.
- [ ] Ability to define and include multiple Vector configuration files (unrelate, but worth mentioning)

**Comment by Jeffail** [comment]
I like the private components idea.

Regarding `pipelines` versus a new transform, I'm suggesting a transform here because the `pipelines` concept doesn't gell well with copies over references. Imagine our root document where we wish to import some pipeline `foo` defined elsewhere. A `compose` transform can be dropped into the middle of an existing topology:

```toml
[sources.someplace]
  type = "bar"

[transforms.my_transform]
  inputs = [ "someplace" ]
  type = "thing"

[transforms.composed]
  inputs = [ "my_transform" ]
  type = "compose"
  children = [ "foo" ]

[sinks.somewhere]
  inputs = [ "composed" ]
  type = "baz"
```

Because it has a clear distinction between its own `inputs`, which are references to global components with state, versus the composed children which are isolated copies. By contrast, in a pipeline example:

```toml
[sources.someplace]
  type = "bar"

[transforms.my_transform]
  inputs = [ "someplace" ]
  type = "thing"

pipelines.composed = [ "my_transform", "foo" ]

[sinks.somewhere]
  inputs = [ "composed" ]
  type = "baz"
```

How do we know which transforms should be copies? Should we assume that the first transform is a reference, the others copies? In that case not only is it awkward to document and understand but it also means we're permanently limited to only one input.

We can solve this with syntax, maybe start bringing in pointer semantics:

```yaml
pipelines.composed = [
  "&my_transform", # Reference
  "foo"            # Copy
]
```

But that's getting a little gross for a configuration spec in my opinion. It also doesn't solve >1 inputs without adding another syntax.

**Comment by binarylogic** [comment]
I think we're thinking about pipelines differently. In your above example:

```toml
[sources.someplace]
  type = "bar"

[transforms.my_transform]
  inputs = [ "someplace" ]
  type = "thing"

pipelines.composed = [ "my_transform", "foo" ]

[sinks.somewhere]
  inputs = [ "composed" ]
  type = "baz"
```

I see 2 incomplete pipelines:

1. `someplace` -> `my_transform`
2. `composed` -> `somewhere`

The fact that `my_transform` is the first in the pipeline does not mean it is connected. It simply means that it contains a copy of the `my_transform` component.

To complete this, I would either need to:

1.  Add a source to the pipeline:
2.  Define a separate pipeline with a source.

Here are 3 examples:

### 1. Add a source to the pipeline

```toml
pipelines.composed = [ "someplace", "my_transform", "foo" ]

[sinks.somewhere]
  inputs = [ "composed" ]
  type = "baz"
```

### 2. Define an entire pipeline

```toml
pipelines.composed = [ "someplace", "my_transform", "foo", "somewhere" ]

[sinks.somewhere]
  type = "baz"
```

### 3. Define a separate pipeline


```toml
pipelines.composed = [ "my_transform", "foo", "somewhere" ]
pipelines.complete = [ "someplace", "composed" ]

[sinks.somewhere]
  inputs = ["complete"]
  type = "baz"
```

Ja feel?

**Comment by Jeffail** [comment]
Okay in that case we need to solve another problem. How do we define something like this:

```toml
[sources.someplace]
  type = "bar"

[sources.another_place]
  type = "bar"

[transforms.my_transform]
  inputs = [ "someplace" ]
  type = "thing"

[transforms.composed]
  inputs = [ "another_place", "my_transform" ]
  type = "compose"
  children = [ "foo" ]

[sinks.somewhere]
  inputs = [ "composed" ]
  type = "baz"
```

With pipelines? We have two inputs, where we want to feed one into a transform and the output of that transform and the other input into our externally defined pipeline.

**Comment by binarylogic** [comment]
That's a good point. I can think of three ways:

### 1. Coupling sources

```toml
pipelines.composed = [ ["someplace", "another_place"], "my_transform", "foo" ]
```

### 2. Disparate pipelines sources

```toml
pipelines.composed = [ "my_transform", "foo" ]
pipelines.pipeline1 = [ "someplace", "composed" ]
pipelines.pipeline2 = [ "another_place", "composed" ]
```

### 3. Support a pipelines directive in the source itself (I dislike this the most)

```toml
[sources.someplace]
  type = "file"
  pipelines = ["composed"]
```

---

The more we work through this, the more I realize these features aren't mutually exclusive. My only concern is simplicity, having both could be confusing to users. And I view pipelines as a way to take a step back and define a full end-to-end pipeline. Yes, it could be used to partially define a pipeline, but the end-goal is to create a short, concise pipelines directive that makes it easy to understand the topology.

**Comment by Jeffail** [comment]
I've added the next proposal which is a continuation of pipelines: https://github.com/timberio/vector/issues/1679, let's move the discussion of pipelines to that ticket and keep this one specifically about the compose transform.

**Comment by Hoverbear** [comment]
After some discussion this will be the basis of the #750 (shape transform) since it will be a compose of add/remove/rename. So I'll be closing that one and directing conversation here.

**Comment by binarylogic** [comment]
@Hoverbear I've assigned this to you and moved it to your project since it's related.

**Comment by binarylogic** [comment]
I'm closing this issue since it was opened at time where we were working out a variety of issues with Vector:

1. The pipeline syntax, which is being addressed in https://github.com/timberio/vector/issues/1327.
2. Mapping syntax, which is being addressed in https://github.com/timberio/vector/pull/3134.
